pkg/util/filesystem/concurrent_write.go

```diff
@@ -0,0 +1,158 @@
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package filesystem
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/gofrs/flock"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// ArtifactBuilder is a generic interface for building, serializing, and deserializing artifacts.
+// The type parameter T represents the in-memory type of the artifact.
+type ArtifactBuilder[T any] interface {
+	Generate() (T, error)
+	Serialize(T) ([]byte, error)
+	Deserialize([]byte) (T, error)
+}
+
+// FetchArtifact attempts to load an artifact using the provided factory.
+// If the artifact does not exist, it return an error.
+func FetchArtifact[T any](location string, factory ArtifactBuilder[T]) (T, error) {
+	var zero T
+
+	// Read the artifact
+	content, err := os.ReadFile(location)
+	if err != nil {
+		return zero, fmt.Errorf("unable to read artifact: %s", err.Error())
+	}
+
+	// Try to load artifact
+	res, err := factory.Deserialize(content)
+	return res, err
+}
+
+// FetchOrCreateArtifact attempts to load an artifact using the provided factory.
+// If the artifact does not exist, it generates a new one, stores it, and returns it.
+//
+// The function first tries to load the artifact using the factory's location.
+// If loading fails, it generates a temporary artifact and attempts to acquire a file lock.
+// When the lock is acquired, the function checks if another process has already created the artifact.
+// If not, it moves the temporary artifact to its final location.
+//
+// The function will repeatedly try to acquire the lock until the context is canceled or the lock is acquired.
+//
+// This function is thread-safe and non-blocking.
+func FetchOrCreateArtifact[T any](ctx context.Context, location string, factory ArtifactBuilder[T]) (T, error) {
+	var zero T
+
+	res, err := FetchArtifact(location, factory)
+	if err == nil {
+		return res, nil
+	}
+
+	// Happy path is to be able to generate and store artifact
+	// We prefer generating a temporary artifact and moving it to its final location
+	// to avoid having a half written artifact in case of a failure
+	createdArtifact, tmpLocation, err := generateTmpArtifact(location, factory)
+	if err != nil {
+		return zero, fmt.Errorf("unable to generate temporary artifact: %v", err.Error())
+	}
+
+	fileLock := flock.New(location + ".lock")
+	defer func() {
+		log.Debugf("releasing lock for file %v", location)
+		err := fileLock.Unlock()
+		// We don't want to remove the lock file here, as it may be used by another process
+		if err != nil {
+			log.Warnf("unable to release lock: %v", err)
+		}
+	}()
+
+	// trying to lock artifact file
+	locked, err := fileLock.TryLockContext(ctx, 100*time.Millisecond)
+
+	if err != nil || !locked {
+		if err == nil {
+			err = fmt.Errorf("unknown error")
+		}
+		if !locked {
```

Comment:
I think here you'll always be unlocked (in case of error the lock should stay open)